File: src/Pacman/Map.cpp
#include "AphMain.h"
#include "Map.h"


MapTile& Map::GetTileByFunction(int specialFunction) {
	for (auto& til : blocks) {
		if (til.second.specialFunction == specialFunction) {
			return til.second;
		}
	}
	throw std::invalid_argument("Tile not found");
}


bool Map::CanGoAlongDirection(Direction dir, int x, int y) {
	switch (dir) {
	case Direction::LEFT: return CanGoLeft(x, y);
	case Direction::RIGHT: return CanGoRight(x, y);
	case Direction::UP: return CanGoUp(x, y);
	case Direction::DOWN: return CanGoDown(x, y);
	}
}


Direction Map::GetOppositeDirection(Direction dir, int x, int y) {
	switch (dir) {
	case Direction::LEFT: return CanGoRight(x, y) ? Direction::RIGHT : dir;
	case Direction::RIGHT: return CanGoLeft(x, y) ? Direction::LEFT : dir;
	case Direction::UP:return CanGoDown(x, y) ? Direction::DOWN : dir;
	case Direction::DOWN: return CanGoUp(x, y) ? Direction::UP : dir;
	}
}

bool Map::CanGoLeft(int x, int y) {
	auto& currentTile = GetTile(x, y);
	return currentTile.CanGoLeft();
}

bool Map::CanGoRight(int x, int y) {
	auto& currentTile = GetTile(x, y);
	return currentTile.CanGoRight();
}

bool Map::CanGoUp(int x, int y) {
	auto& currentTile = GetTile(x, y);
	auto& upTile = GetTile(x, y - 1);

	// hardcoded conditions -> we assume that the gate isn't accessible neither from left nor right and the spawner (spider gate) isn't accessible from top
	return currentTile.CanGoUp() &&
		upTile.specialFunction != FUNCTION_SPAWNER && (upTile.specialFunction != FUNCTION_GATE || upTile.state == STATE_GATE_OPENED); // hardcoded conditions
}

bool Map::CanGoDown(int x, int y) {
	auto& currentTile = GetTile(x, y);
	auto& downTile = GetTile(x, y + 1);

	// hardcoded conditions -> we assume that the gate isn't accessible neither from left nor right
	return currentTile.CanGoDown()
		&& (downTile.specialFunction != FUNCTION_GATE || downTile.state == STATE_GATE_OPENED);
}

File: src/games/pacman/map.ts
import { SpecFunctions, GateState, STATE_DEFAULT } from './constants';
import { Direction } from './direction';
import { Vector } from '../../../libs/pixi-component';

/**
 * Map cell
 */
export class MapTile {
  pos: Vector; // position in the grid
  code: number; // binary code for walking possibilities
  specialFunction: number; // special func (tunnel, gate, spawner,...)
  state = 0; // dynamic state, can be set during the game

  canGoLeft() {
    return (this.code & 0b0010) === 0b0010;
  }

  canGoRight() {
    return (this.code & 0b0001) === 0b0001;
  }

  canGoUp() {
    return (this.code & 0b1000) === 0b1000;
  }

  canGoDown() {
    return (this.code & 0b0100) === 0b0100;
  }
}

export class GameMap {
  private blocks = new Map<number, MapTile>();
  // helper for faster search of tiles by their functions
  private functions = new Map<number, MapTile[]>();
  private _gridWidth: number;
  private _gridHeight: number;

  constructor(blocks: Map<number, MapTile>, gridWidth: number, gridHeight: number) {
    this.blocks = blocks;
    this._gridWidth = gridWidth;
    this._gridHeight = gridHeight;
  }

  get gridWidth() {
    return this._gridWidth;
  }

  get gridHeight() {
    return this._gridHeight;
  }

  init() {
    this.functions.clear();
    for(let [,tile] of this.blocks) {
      tile.state = STATE_DEFAULT; // reset all states
      if(tile.specialFunction !== SpecFunctions.NONE) {
        if(!this.functions.has(tile.specialFunction)) {
          this.functions.set(tile.specialFunction, []);
        }
        this.functions.get(tile.specialFunction).push(tile);
      }
    }
  }

  getTileByFunction(specialFunction: number): MapTile {
    if(this.functions.has(specialFunction)) {
      return this.functions.get(specialFunction)[0];
    }
    return null;
  }

  getTilesByFunction(specialFunction: number): MapTile[] {
    if(this.functions.has(specialFunction)) {
      return [...this.functions.get(specialFunction)]; // make copy
    }
    return null;
  }

  getTile(x: number, y: number): MapTile {
    return this.blocks.get(y*this.gridWidth + x) as MapTile;
  }

  canGoAlongDirection(dir: Direction, pos: Vector) {
    switch(dir) {
      case Direction.LEFT: return this.canGoLeft(pos);
      case Direction.RIGHT: return this.canGoRight(pos);
      case Direction.UP: return this.canGoUp(pos);
      case Direction.DOWN: return this.canGoDown(pos);
    }
  }

  canGoLeft(pos: Vector): boolean {
    // TODO if the gate could be accessible from left or right, add the same condition as for canGoUp/down
    return this.getTile(pos.x, pos.y).canGoLeft();
  }

  canGoRight(pos: Vector): boolean {
    // TODO if the gate could be accessible from left or right, add the same condition as for canGoUp/down
    return this.getTile(pos.x, pos.y).canGoRight();
  }

  canGoUp(pos: Vector): boolean {
      let currentTile = this.getTile(pos.x, pos.y);
      let upTile = this.getTile(pos.x, pos.y - 1);
      return currentTile.canGoUp() && (upTile.specialFunction !== SpecFunctions.GATE || upTile.state === GateState.OPEN);
  }

  canGoDown(pos: Vector): boolean {
      let currentTile = this.getTile(pos.x, pos.y);
      let downTile = this.getTile(pos.x, pos.y + 1);
      return currentTile.canGoDown() && (downTile.specialFunction !== SpecFunctions.GATE || downTile.state === GateState.OPEN);
  }
}